go/vt/servenv/embedder.go

```diff
@@ -0,0 +1,444 @@
+		delete(vmap.vars, e.name)
+		vmap.mu.Unlock()
+	}
+
+	if e.sp != nil {
+		e.sp.reset()
+	}
+}
+
+// findOrCreateEmbedder returns an existing embedder or
+// creates a new one.
+func findOrCreateEmbedder(instanceName string) *embedder {
+	if e, ok := embeds[instanceName]; ok {
+		return e
+	}
+	e := &embedder{
+		name:        instanceName,
+		prefix:      "",
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
+		e.sp = newStatusPage(instanceName)
+	}
+	embeds[instanceName] = e
+	return e
+}
+
+// URLPrefix returns the URL prefix for all the embedder.
+func URLPrefix(instanceName string) string {
+	// There are two other places where this logic is duplicated:
+	// status.go and go/vt/vtgate/discovery/healthcheck.go.
+	if instanceName == "" {
+		return ""
+	}
+	return "/" + instanceName
+}
+
+// HandleFunc sets or overwrites the handler for url. If embedder has a name,
+// url remapped from /path to /name/path. If name is empty, the request
+// is passed through to http.HandleFunc.
+func HandleFunc(instanceName, url string, f func(w http.ResponseWriter, r *http.Request)) {
+	if instanceName == "" {
+		http.HandleFunc(url, f)
+		return
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+
+	hf, ok := e.handleFuncs[url]
+	if ok {
+		hf.Set(f)
+		return
+	}
+	hf = &handleFunc{f: f}
+	e.handleFuncs[url] = hf
+
+	http.HandleFunc(URLPrefix(e.name)+url, func(w http.ResponseWriter, r *http.Request) {
+		if f := hf.Get(); f != nil {
+			f(w, r)
+		}
+	})
+}
+
+// AddInstanceStatusPart adds a status part to the status page. If instance has a name,
+// the part is added to a url named /name/debug/status. Otherwise, it's /debug/status.
+func AddInstanceStatusPart(instanceName, banner, frag string, f func() interface{}) {
+	if instanceName == "" {
+		AddStatusPart(banner, frag, f)
+		return
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+	e.sp.addStatusPart(banner, frag, f)
+}
+
+// NewCountersFuncWithMultiLabels creates a name-spaced equivalent for stats.NewCountersFuncWithMultiLabels.
+func NewCountersFuncWithMultiLabels(instanceName, name, help string, labels []string, f func() map[string]int64) *stats.CountersFuncWithMultiLabels {
+	// If instanceName is empty, it's a pass-through.
+	// If name is empty, it's an unexported var.
+	if instanceName == "" || name == "" {
+		return stats.NewCountersFuncWithMultiLabels(name, help, labels, f)
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+
+	if vmap, ok := globalStatVars[name]; ok {
+		vmap.Set(e.name, f)
+		return stats.NewCountersFuncWithMultiLabels("", help, labels, f)
+	}
+	vmap := &varMap{vars: map[string]func() map[string]int64{e.name: f}}
+	globalStatVars[name] = vmap
+
+	newlabels := append(append(make([]string, 0, len(labels)+1), e.prefix), labels...)
+	_ = stats.NewCountersFuncWithMultiLabels(e.prefix+name, help, newlabels, func() map[string]int64 {
```

Comment:
It seems wrong to me to include `e.prefix` in the name of the metric *and* in a label. Is this actually correct?